src/io/dataset.cpp

```diff
@@ -356,6 +369,15 @@ void Dataset::ResetConfig(const char* parameters) {
   if (param.count("sparse_threshold") && io_config.sparse_threshold != sparse_threshold_) {
     Log::Warning("Cannot change sparse_threshold after constructed Dataset handle.");
   }
+  if (param.count("forcedbins_filename")) {
+    /* Since the dataset is already constructed we don't know which bins are categorical.
+    Therefore read forced bins assuming no categorical features, and warn if not the same as original. */
+    std::vector<std::vector<double>> config_bounds = Dataset::GetForcedBins(io_config.forcedbins_filename, 
+                                                                            num_total_features_, std::unordered_set<int>());
```

Comment:
categorical feature is needed here?

Reply:
In this case the dataset is already constructed, and I think there is no way to know which features are categorical (because this information is only stored in the `DatasetLoader` object, not in `Dataset`). Therefore I think the best we can do is get the forced bins assuming no categorical features, and warn the user if these are different from the existing bins.

Reply:
actually, you can call this to get the bin mapper: https://github.com/microsoft/LightGBM/blob/f1a1486929f8e981c44cba45132f5b4929453f8a/include/LightGBM/dataset.h#L476-L480 . And I think config_bounds will always no equal to forced_bin_bounds_ when categorical feature is set.

Reply:
Thanks, I changed it to get the categorical features from FeatureBinMapper. But you're right: when the json file contains forced bins for categorical features it always gives the warning, even if forced bins are unchanged.